### ISSUE TITLE
NodeIterator's pre-removing steps does not match browsers

### ISSUE THREAD

**Issue by TRowbotham** [whatwg/dom#907]
The following is a reduced example from the [wpt](https://github.com/web-platform-tests/wpt/blob/master/dom/traversal/NodeIterator-removal.html), where "paras[0]" is the node being removed and "paras[0].firstChild" is the node used to create the NodeIterator.
```js
const div = document.createElement('div');
const p = document.createElement('p');
const t = p.appendChild(document.createTextNode('Hello'));
div.appendChild(p);

const iter = document.createNodeIterator(t);
div.removeChild(p);
```

The html structure is:
```html
<div>
  <p>Hello</p>
</div>
```

The NodeIterator's pre-removing steps:
1. The node being removed (`p`) is an inclusive ancestor of node iterator's reference (`t`), so this passes.
2. The node iterator's `pointer before reference` starts out as `true`, so this passes.
    1. There is no following node that is an inclusive descendant of node iterator's `root` (`t`) that is also not an inclusive descendant of the node being removed (`p`), which means next is null.
    2. Next is null, so we move to the next step.
    3. **Because next is null, node iterator's `pointer before reference` gets set to false.**
3. **The node being removed (`p`) has no previous sibling, so node iterator's `reference` gets set to the removed node's (`p`) parent, which is `div`.**

Following the spec, the NodeIterator gets changed to the following:
```js
NodeIterator {
  root: t,
  referenceNode: div,
  pointerBeforeReferenceNode: false
}
```

Browsers, however, seem to leave the NodeIterator unchanged, which matches the test expectations:
```js
NodeIterator {
  root: t,
  referenceNode: t,
  pointerBeforeReferenceNode: true
}
```

**Comment by annevk** [comment]
Looking at https://searchfox.org/mozilla-central/source/dom/base/NodeIterator.cpp#63 suggests the "not" in the first step is wrong. What's also interesting is that the reference apparently can be null in Firefox, though the specification has no such path.

**Comment by TRowbotham** [comment]
Wouldn't removing the "not" in the first step defeat the whole purpose of these steps? If you remove it, then we return early when toBeRemovedNode is an inclusive ancestor of reference, which means reference would still point to a node that has now been removed from the tree, right?

**Comment by annevk** [comment]
That's correct, but the steps would still run when that's not the case. I don't think it's always possible to end up with a sensible reference.

And a good first step would be to match browsers here, before trying to attempt to improve the situation for certain edge cases, no?

**Comment by TRowbotham** [comment]
Agreed, matching existing behavior first seems logical.

**Comment by annevk** [comment]
@TRowbotham to confirm, if you make that change in your implementation, do the other tests still pass?

**Comment by TRowbotham** [comment]
To be clear, I am seeing 7 failures with the current behavior. There could potentially be more, but my test harness bails when an assertion fails. My example in the first post is just the first one that fails.

The following pairs are failing, where the first item is the node being removed and the second is the Node Iterator's root.
1. ["paras[0]", "paras[0].firstChild"]
2. ["foreignPara1", "foreignPara1.firstChild"]
3. ["detachedPara1", "detachedPara1.firstChild"]
4. ["xmlElement", "xmlTextNode"]
5. ["paras[1]", "paras[1].firstChild"]
6. ["detachedPara2", "detachedPara2.firstChild"]
7. ["testDiv", "paras[0]"]

If I remove the "not" from the first step, the failure count jumps to 22.

**Comment by annevk** [comment]
That's not great, I guess this requires a lot more investigative work then.

**Comment by sideshowbarker** [comment]
Note that Ladybird has implemented the relevant spec requirements for this as currently written, but isn’t passing the relevant WPT tests at https://wpt.fyi/results/dom/traversal/NodeIterator-removal.html?label=master&product=chrome%5Bexperimental%5D&product=firefox%5Bexperimental%5D&product=safari%5Bexperimental%5D&product=ladybird&aligned due to the problem described in the issue description. (That is, due to the WPT tests apparently having been written to match what other engines have implemented — but not written to match the actual spec.)

**Comment by bptato** [comment]
To my understanding, there are three cases to consider:

1. if toBeRemovedNode is not an inclusive ancestor of referenceNode, then referenceNode is unaffected by the removal, so bail.  e.g.
	* root
		* toBeRemovedNode
		* referenceNode
2. if toBeRemovedNode is an inclusive ancestor of root, then the NodeIterator will be completely detached from the tree anyway, so bail.  e.g.
	* toBeRemovedNode
		* root
			* referenceNode
3. if toBeRemovedNode is a descendant of root, but also an inclusive ancestor of referenceNode, *then* set referenceNode to something that isn't an inclusive descendant of toBeRemovedNode.  e.g.
	* root
		* toBeRemovedNode
			* referenceNode
		* someOtherNode (will become referenceNode)

As it stands, the spec does not fully cover case 2; I believe it should read:

> 1. If toBeRemovedNode is not an inclusive ancestor of nodeIterator’s reference, or toBeRemovedNode is **an inclusive ancestor of** nodeIterator’s root, then return.

(At least my naive implementation passes the aforementioned WPT tests with the above change.
Looking at the Gecko link, it apparently invokes the removing steps from "ContentWillBeRemoved", so maybe case 2 is implicitly fulfilled there.)

**Comment by annevk** [comment]
@sideshowbarker @bptato I created #1477 for this and AI also managed to find an interoperability issue between Gecko and WebKit (/Chromium): https://github.com/web-platform-tests/wpt/pull/60997.
